Removed bl typings as integrated in package starting from v5.1.0

alexandercerutti opened the pull request:
Closes #62869 

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

alexandercerutti commented:
I just opened the PR to [DefinitelyTyped-tools](https://github.com/microsoft/DefinitelyTyped-tools): https://github.com/microsoft/DefinitelyTyped-tools/pull/547

alexandercerutti commented:
The PR has been merged, but I honestly don't understand what I have to change now 🤔 Any hint, anyone?

benallfree commented:
@alexandercerutti I'm not sure that removing this package is the best idea. What about people using previous versions of `bl` that depend upon this?

benallfree reviewed (changes requested):
Move `msgpack5` to a separate PR, I think that might be messing up CI because the `bl` type maintainers have no authority to approve `msgpack5` changes and vice versa.

alexandercerutti commented:
@benallfree Thanks for your comment! I just followed what it is written in the Readme. Removal is just a deprecation actually. It makes no sense to have bl's declarationa in DT when it ship it already.

I added msgpack5 in here because it was written in the docs.

I think having it in a different PR would break them in the same way 🤔 I don't know if there's an order that should be respected for this.

benallfree commented:
@alexandercerutti I've now read the instructions more carefully and agree with all your steps. `@types/bl@xx` will remain published in `npm` as you say, so there is no need to maintain it here any longer.

I believe now that https://github.com/microsoft/DefinitelyTyped-tools/blob/master/packages/dtslint/docs/no-outside-dependencies.md is the reason CI is failing. Let me try an edit to see.

alexandercerutti commented:
@benallfree what should we put instead of that? I mean, `bl` is used there. 🤔

benallfree commented:
I think all types are in the search path already. Try removing it and see if tests pass :)

benallfree commented:
I'll see if I can find some time to check it out.

alexandercerutti commented:
So, in order to recap, it seems that `msgpack5` requires `safe-buffer` dependency. It actually uses `safe-buffer` on runtime, but it doesn't in types. Also, adding `safe-buffer` to `package.json`, seems not to solve the issue (even if it should, according to error messages).

I also added it to declarations and replaced the `Buffer` import with the one provided within `safe-buffer`, but no changes.

So, I don't have any idea right now how to solve this.

benallfree commented:
@alexandercerutti I have a passing PR here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62869

If you want to update this PR to match so you get contrib credit, please go ahead. Otherwise you can close this one and we can work this issue from my PR 👍

alexandercerutti commented:
Hey @benallfree, for me it is the same thing actually! I just took charge of the deprecation on DT types. I still honestly don't understand why you don't have tests issue while I do 🤔 Was it due to `readable-stream`?!

benallfree commented:
@alexandercerutti Yes, `readable-streams` was causing test failures by using an unspecified version of `safe-buffer`, but this only became a problem after changing `msgpack5`'s dependencies.

If you can't get this branch working, we can proceed with my PR. I'm not sure how to give you credit though.

DangerBotOSS commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## readable-stream ([<kbd>unpkg</kbd>](https://unpkg.com/browse/readable-stream@latest/))
was missing the following properties:
1. from
1. finished
1. pipeline
1. once
1. on

<details>
<summary>as well as these 6 other properties...</summary>
<p>listenerCount, errorMonitor, captureRejectionSymbol, captureRejections, defaultMaxListeners, EventEmitter</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2ea47cf1c7b90b32afc4564276512999f641b6e6
</p>

benallfree commented:
@alexandercerutti Please take a look at https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62783#issuecomment-1292353562 to make sure we are getting typings for all the props.

alexandercerutti commented:
@benallfree oh so you mean we should update also `readable-stream`? Is it in the scope of this PR?

benallfree commented:
@alexandercerutti That is my opinion, yes. `readable-stream` is at v5.x now, and the typings are for v2.3 so they are very out of date. As you noticed, `readable-stream` no longer uses `safe-buffer`. I assume there are some other changes too. I think the typings should be updated to v5.x as part of this PR. But it's up to you. I'm just not sure what the best way is to fix all this.

alexandercerutti commented:
@benallfree I think that readable-stream (which is v4.x, not v5) is a pretty critical package. I mean, I know that its interface is the same as NodeJS streams (but with different versioning), but the fact that no one kept updating it, means that one should:

 - Verify the current state of the types (if they actually respect v2.3)
 - Get back to Node v9 and Node v10 and check all the breaking changes (they are actually reported in the package readme)
 - Create a package for v3.x
 - Start from Node v12 up to v18, check all the changes to the API
 - Create a package for v4.x
 
I mean, this is a huge and time-consuming job, don't you agree? It should be done and I might do it, but I wouldn't do it here. That's not a very-blocking thing.

One thing that I'd do, instead, is to update the version of readable-stream types to tell they are for `>= v2.3.7` (as per what you said [in your comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62869#discussion_r1004937189) on #62869), instead of `v2.3`. I don't know if that's possible. I'm going to take a couple of attempts.

alexandercerutti commented:
@benallfree I'm a bit in a contrast here for a few reasons:

 - `readable-stream@v2.3.7` is the last v2.x and the last to use `readable-stream`, so removing the dependency is correct for the next major versions, not others v2.x (so we should);
 - To have a correct update path, we should:
   - Verify the current state of the types (if they actually respect v2.3)
   - Get back to Node v9 and Node v10 and check all the breaking changes (they are actually reported in the package readme)
   - Create a package for v3.x
   - Start from Node v12 up to v18 and check all the breaking changes to the API
   - Create a package for v4.x
 
I mean, this is a huge and time-consuming job, don't you agree? It should be done and I might do it, but I definitely underestimated the work that had to be done.

benallfree commented:
I agree with your analysis @alexandercerutti. @jakebailey any thoughts to add? 

@alexandercerutti my PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62869 fixes it by explicitly naming the `safe-buffer` version. What if we just used my approach and bumped `@types/readable-streams` to `2.3.1` or something? Leaving the big job for someone more connected to the `readable-streams` project?

alexandercerutti commented:
I think that having a specific version of safe-buffer inside the `readable-stream`'s `package.json` is the best way rn. I'd copy npm semver selector `~5.1.1` as a version. By doing this, or with `^5.1.2` (but not `^5.1.1`, and I don't understand why), it seems to keep working.

benallfree commented on types/msgpack5/package.json:
```diff
@@ -1,6 +1,6 @@
     "private": true,
     "dependencies": {
-        "safe-buffer": "*"
+        "bl": ">=5.1.0"
```
This should be `~5.1.2` because `5.1.2` is the first version of `bl` that contains its own typings.

benallfree commented on types/readable-stream/index.d.ts:
```diff
@@ -7,8 +7,6 @@
 
```
Do we know how to bump the version of `@types/readable-stream`? Is it this:

```
// Type definitions for readable-stream 2.3.1
```

alexandercerutti replied:
I think that bumping is automatic through typescript-bot when it is going to release this PR. Major and minor didn't change, so I guess it will publish a patch.

alexandercerutti commented on types/readable-stream/index.d.ts:
```diff
@@ -158,7 +156,7 @@ declare namespace _Readable {
     }
 
-    interface BufferList<D extends SafeBuffer.Buffer = SafeBuffer.Buffer> {
+    interface BufferList<D extends Buffer = Buffer> {
```
This is getting reverted if we proceed with the way you proposed. Otherwise, I'll have to check exactly.

benallfree replied:
👍 Right, sorry. Yes I do think we should proceed as agreed by keeping `safe-buffer` in there.

alexandercerutti commented on types/readable-stream/index.d.ts:
```diff
@@ -80,53 +78,53 @@ declare class _Readable implements _IReadable {
-    addListener(event: "end", listener: () => void): this;
-    addListener(event: "readable", listener: () => void): this;
-    addListener(event: "error", listener: (err: Error) => void): this;
+    addListener(event: 'close', listener: () => void): this;
```
DT has prettier. I have to run it.

benallfree replied:
I think when you revert it won't matter anyway 👍

alexandercerutti replied:
Prettier changed many other things ahahah

benallfree replied:
Yeah just revert it. This file doesn’t need to be touched.

benallfree reviewed (approved):
Awesome!!! Great job @alexandercerutti, this was a tough one!

alexandercerutti commented:
Alleluja! 😂 We did it!

TeamworkGuy2 commented on types/readable-stream/package.json:
```diff
@@ -1,6 +1,6 @@
     "private": true,
     "dependencies": {
-        "safe-buffer": "*"
+        "safe-buffer": "~5.1.1"
```
TS Types were added to `safe-buffer` in version [5.1.2](https://github.com/feross/safe-buffer/commit/9bcb1129bfadc78401cf5d8968762d8fa35304f8), should we pin that version?
```suggestion
        "safe-buffer": "~5.1.2"
```

alexandercerutti replied:
@TeamworkGuy2 Uhm, that's a good point, but I set `~` to be coherent with [`readable-streaming@2.3.7`'s `package.json`](https://github.com/nodejs/readable-stream/blob/v2.3.7/package.json#L11), which tells to the package manager to look only at patches. So I don't think this is an issue. 5.1.2 will still get targeted.

Thank you for your review.

jakebailey reviewed (approved):
LGTM; the next step is to update readable stream's major version, right?

alexandercerutti commented:
@jakebailey exactly. It is going to be kind of a mess 😄

alexandercerutti commented:
Ready to merge
